### PR TITLE
Initial demo version of libraries API with example implementations

### PR DESCRIPTION
**Pull request by Moleus**
This pull request adds the initial demo version of the Libraries API. The API comprises of functions that enable users to interact with the motor and structures that represent its connection type and state. 

_The Protocol Library_ is a self-contained piece of code that implements all the logic required to communicate with a motor via a specific protocol, such as `UART` or `GPIO`. 

Library headers are divided into two parts: internal and external (API). The `MotorAttributes` structure is opaque, which hides most of the implementation details. 

Furthermore, two files, namely `lib-common-example.c` and `lib-gpio-example.c`, are included in this version. These files serve as templates for future implementations. The former is intended to be shared between different libraries, while the latter implements the functions in a protocol-specific manner.

### CODE REVIEW

File: src/proto-libs/internal.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    /* bool max_triggered; */
+    /* bool min_triggered; */
+} MotorState;
+
+
+/* Contains attributes of a single motor.
+* Filled at the initialization phase.
+*/
+struct MotorAttributes {
+    MovementResult (* proto_specific_move)(MotorState* state, int steps);
```
Comment (filberol) [inline src/proto-libs/internal.h:32]: typedef

File: src/proto-libs/api.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+void motor_set_min_enstop(MotorAttributes* attributes, int min_pin);
+
+// TODO: decide where to start counting the steps. From the left or from center?
+/* sets max/min positions for the motor */
+void motor_set_min_softlimit(MotorAttributes* attributes, int min_steps);
+
+void motor_set_max_softlimit(MotorAttributes* attributes, int max_steps);
+
+/* inverts default rotation direction */
+void motor_invert_direction(MotorAttributes* attributes, bool invert);
```
Comment (filberol) [inline src/proto-libs/api.h:73]: motor_set_axis_inversion(MotorAttributes* attributes, bool invert);

File: src/proto-libs/internal.h (hunk 1 above)
Comment (filberol) [inline src/proto-libs/internal.h:32]: Typedef?

File: src/proto-libs/api.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+/* Represents how the stepper is connected to the main board */
+typedef enum {
+    GPIO = 0,
+    SPI,
+    I2C
+} CommunicationType;
+
+typedef struct {
+    int pins[4];
```
Comment (filberol) [inline src/proto-libs/api.h:17]: Use uintxx_t?

File: src/proto-libs/api.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+} ConnectionAttributes;
+
+/* Stores data required to communicate with a motor/extension-board */
+typedef struct {
+    CommunicationType communicationType;
+    ConnectionAttributes connectionAttributes;
+} MotorConnection;
+
+/* Represent the status of movement invocation */
+typedef enum {
```
Comment (filberol) [inline src/proto-libs/api.h:42]: Note: function invocation is blocking, process monitoring in other layer

**Review by filberol — APPROVED** [review]

File: src/proto-libs/api.h (hunk 2 above)
Comment (Moleus) [reply src/proto-libs/api.h]: fixed in [959b203](https://github.com/ThatDraenGuy/motorland/pull/5/commits/959b203999521a3d2ddddeba920d73808577f814)

File: src/proto-libs/internal.h (hunk 1 above)
Comment (Moleus) [reply src/proto-libs/internal.h]: fixed in [83a4b31](https://github.com/ThatDraenGuy/motorland/pull/5/commits/83a4b31c987b41c04cb391f9750f26b95757db58)

File: src/proto-libs/api.h (hunk 3 above)
Comment (Moleus) [reply src/proto-libs/api.h]: Replaced all integer types with exact-width ones [8ff2d39](https://github.com/ThatDraenGuy/motorland/pull/5/commits/8ff2d39f24aa77c9be45df3cf0d8aadbb844f471)

**Pull request closed without merging**